File: client-sdk/python/examples/03-prompt-attachment.py
"""Prompt an agent with a file attached.

Mirrors ``examples/03-prompt-attachment.ts`` — demonstrates the boss's
pitch: ``prompt("describe this photo", attachments=[...])``. §5.4
validation runs locally BEFORE any NATS traffic when the agent's prompt
endpoint declared ``attachments_ok: false`` or ``max_payload`` is
exceeded.
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from examples._connect_cli import add_connection_flags, connect_from_cli
from synadia_ai.agents import (
    Agents,
    Attachment,
    AttachmentsNotSupportedError,
    DiscoverFilter,
    NatsAgentError,
    PayloadTooLargeError,
    Query,
    ResponseChunk,
    StatusChunk,
)


async def main() -> None:
    parser = argparse.ArgumentParser(
        description="Prompt the first discovered agent with a file attachment."
    )
    parser.add_argument("path", help="file to attach")
    parser.add_argument(
        "--prompt",
        default="describe this photo",
        help="prompt text (default: 'describe this photo')",
    )
    parser.add_argument(
        "--session",
        default=None,
        metavar="NAME",
        help=(
            "Select the agent whose `session_name` matches NAME — i.e. the "
            "5th subject token (v0.3). When omitted, the first discovered "
            "agent is used."
        ),
    )
    add_connection_flags(parser)
    args = parser.parse_args()

    attachment_path = Path(args.path)
    if not attachment_path.is_file():
        print(f"no such file: {attachment_path}", file=sys.stderr)
        sys.exit(1)

    nc = await connect_from_cli(args)
    agents = Agents(nc=nc)
    try:
        filt = DiscoverFilter(session_name=args.session) if args.session else None
        found = await agents.discover(filter=filt)
        if not found:
            print("no agents reachable — start the reference agent first", file=sys.stderr)
            sys.exit(2)

        chosen = found[0]
        ep = chosen.prompt_endpoint
        mp = ep.max_payload_bytes if ep.max_payload_bytes is not None else "unspecified"
        ao = ep.attachments_ok if ep.attachments_ok is not None else "unspecified"
        print(
            f"prompting {chosen.agent}/{chosen.owner}/{chosen.session_name} "
            f"(max_payload={mp}, attachments_ok={ao})"
        )

        attachment = Attachment.from_path(attachment_path)
        try:
            stream = chosen.prompt(args.prompt, attachments=[attachment])
            async for msg in stream:
                if isinstance(msg, ResponseChunk):
                    sys.stdout.write(msg.text)
                    sys.stdout.flush()
                    if msg.attachments:
                        names = ", ".join(a.filename for a in msg.attachments)
                        print(f"\n  [agent returned {len(msg.attachments)} attachment(s): {names}]")
                elif isinstance(msg, StatusChunk):
                    if msg.status == "done":
                        print("\n[done]")
                elif isinstance(msg, Query):
                    print(f"\n[agent asks: {msg.prompt}]")
                    await msg.reply("ok")
        except AttachmentsNotSupportedError:
            print(
                "\nthis agent does not accept attachments (attachments_ok=false)",
                file=sys.stderr,
            )
            sys.exit(3)
        except PayloadTooLargeError as err:
            print(
                f"\npayload is too large: {err.actual} bytes > agent's {err.limit} byte limit",
                file=sys.stderr,
            )
            sys.exit(4)
    except NatsAgentError as err:
        print(f"demo failed: {type(err).__name__}: {err}", file=sys.stderr)
        sys.exit(99)
    finally:
        await agents.close()
        await nc.close()


if __name__ == "__main__":
    asyncio.run(main())

File: client-sdk/typescript/examples/03-prompt-attachment.ts
// The boss's demo scenario:
//
//     response, error := agent.prompt("describe this photo", WithAttachment("vacation.jpg"));
//
// If the agent declared `attachments_ok: false`, this fails *locally* with
// `AttachmentsNotSupportedError` — no NATS traffic.
// If the prompt + attachment bytes exceed `max_payload`, this fails *locally*
// with `PayloadTooLargeError` — no NATS traffic.
// Otherwise, the prompt is JSON-encoded with a base64 attachment and
// streamed to the agent; the response arrives as typed chunks.
//
// Usage (run the reference agent in one terminal first):
//
//     bun run examples/_run-reference-agent.ts
//     bun run examples/03-prompt-attachment.ts ./vacation.jpg
//
// Or with npm: swap `bun run` for `npx tsx`.

import { argv, exit, stdout } from "node:process";
import { connect as natsConnect } from "@nats-io/transport-node";
import {
  Agents,
  AttachmentsNotSupportedError,
  loadContextOptions,
  NatsAgentError,
  parseNatsUrl,
  PayloadTooLargeError,
  type ResponseAttachment,
} from "@synadia-ai/agents";

async function main(): Promise<void> {
  const attachmentPath = argv[2];
  if (!attachmentPath) {
    console.error("usage: 03-prompt-attachment.ts <path-to-photo>");
    exit(1);
  }

  const opts = process.env["NATS_CONTEXT"]
    ? await loadContextOptions(process.env["NATS_CONTEXT"])
    : process.env["NATS_URL"]
      ? parseNatsUrl(process.env["NATS_URL"])
      : { servers: "nats://127.0.0.1:4222" };
  const nc = await natsConnect(opts);
  const agents = new Agents({ nc });

  try {
    const found = await agents.discover();
    if (found.length === 0) {
      console.error("no agents reachable — start the reference agent first");
      exit(2);
    }

    const chosen = found[0]!;
    console.log(
      `prompting ${chosen.agent}/${chosen.owner}/${chosen.name} (max_payload=${
        chosen.promptEndpoint.maxPayloadBytes ?? "unspecified"
      }, attachments_ok=${chosen.promptEndpoint.attachmentsOk ?? "unspecified"})`,
    );

    try {
      const stream = await chosen.prompt("describe this photo", { attachments: [attachmentPath] });
      for await (const msg of stream) {
        switch (msg.type) {
          case "response":
            stdout.write(msg.text);
            if (msg.attachments) {
              const names = msg.attachments.map((a: ResponseAttachment) => a.filename).join(", ");
              console.log(`\n  [agent returned ${msg.attachments.length} attachment(s): ${names}]`);
            }
            break;
          case "status":
            if (msg.status === "done") console.log("\n[done]");
            break;
          case "query":
            console.log(`\n[agent asks: ${msg.prompt}]`);
            await msg.reply("ok");
            break;
        }
      }
    } catch (err) {
      if (err instanceof AttachmentsNotSupportedError) {
        console.error("\nthis agent does not accept attachments (attachments_ok=false)");
        exit(3);
      }
      if (err instanceof PayloadTooLargeError) {
        console.error(
          `\npayload is too large: ${err.actual} bytes > agent's ${err.limit} byte limit`,
        );
        exit(4);
      }
      throw err;
    }
  } finally {
    await agents.close();
    await nc.close();
  }
}

void main().catch((err: unknown) => {
  if (err instanceof NatsAgentError) {
    console.error("demo failed:", err.name, err.message);
  } else {
    console.error("demo failed:", err);
  }
  exit(99);
});
